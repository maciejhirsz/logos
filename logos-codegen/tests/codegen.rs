use insta::assert_snapshot;
use std::{error::Error, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
#[case("skip_callback")]
#[case("skip_callback_failure")]
#[case("error_callback0")]
#[case("error_callback1")]
#[case("error_callback_failure")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push("codegen");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    let input = std::fs::read_to_string(input)?;

    let generated = logos_codegen::generate(input.parse()?).to_string();

    if cfg!(rust_1_82) {
        assert_snapshot!(format!("{fixture}-1_82"), generated);
    } else {
        assert_snapshot!(format!("{fixture}-pre_1_82"), generated);
    }

    Ok(())
}
