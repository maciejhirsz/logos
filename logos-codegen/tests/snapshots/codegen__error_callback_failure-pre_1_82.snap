---
source: logos-codegen/tests/codegen.rs
expression: generated
---
impl < 's > :: logos :: Logos < 's > for TokenA { type Error = () ; type Extras = () ; type Source = str ; fn lex (lex : & mut :: logos :: Lexer < 's , Self >) { fn _logos_derive_compile_errors () { { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, `dyn`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Expected a named argument at this position\n\nhint: If you are trying to define a callback here use: callback = ...") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Expected: #[logos(error = SomeType)] or #[logos(error(SomeType[, callback))]") } { compile_error ! ("Callback missing a body") } { compile_error ! ("Not a valid callback") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Expected: callback = ...") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Callback has been already set") } { compile_error ! ("Previous callback set here") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } { compile_error ! ("Unknown nested attribute: unknown_attribute\n\nExpected one of: callback") } { compile_error ! ("Error type can be defined only once") } { compile_error ! ("Previous definition here") } } } # [inline] fn make_error (mut lex : & mut :: logos :: Lexer < 's , Self >) { use :: logos :: { Lexer , internal :: LexerInternal } ; let error = callback4 (& mut lex) ; lex . set (Err (error)) ; } }
